File: capacity-club-api/src/domain-modules/exercise-data/model/entity/exercise-data.entity.ts
import { ApiProperty } from '@nestjs/swagger';
import { BaseEntity, Column, Entity, OneToMany, PrimaryColumn } from 'typeorm';
import { ExerciseDataCategory } from '../enum/exercise-data-category.enum';
import { UniqueId, uniqueIdTransformer } from '@common/model/unique-id';
import { Transform } from 'class-transformer';
import { ExerciseTraining } from 'domain-modules/exercise-training/model';

/**
 * Entity representing exercise data.
 * This entity holds the basic information of exercises, which can be reused and adapted by coaches
 * to create specific exercise programs. Coaches can refer to this data to avoid repeating the content
 * of an exercise and adapt details like the number of repetitions in ExerciseTraining.
 */
@Entity()
export class ExerciseData extends BaseEntity {
  /**
   * Unique identifier for the exercise data.
   * Uses a custom transformer for serialization.
   */
  @PrimaryColumn('varchar')
  @Transform(uniqueIdTransformer.to, { toClassOnly: true })
  @Transform(uniqueIdTransformer.from, { toPlainOnly: true })
  exercise_data_id: UniqueId;
  @Column({ nullable: false })
  title: string;
  @Column('text', { nullable: false, array: true, default: [] })
  description: string[];
  @Column('text', { nullable: true, array: true, default: [] })
  categories: ExerciseDataCategory[];
  @Column({ nullable: true })
  video_youtube_link: string;
  @Column({ nullable: true })
  video_asset_link: string;
  @OneToMany(
    () => ExerciseTraining,
    (exerciseTraining) => exerciseTraining.exercise_data,
    {
      cascade: true,
      lazy: true,
      nullable: true,
    },
  )
  @ApiProperty({ type: () => ExerciseTraining, isArray: true })
  exercise_training_list: Promise<ExerciseTraining[]>;
}

File: capacity_club_mobile_app/lib/domain-features/domain/exercise-data/entity/exercise_data_entity.dart
import 'package:capacity_club_mobile_app/core/model/entities/unique_id.dart';
import 'package:capacity_club_mobile_app/domain-features/data/exercise-data/model/enum/exercise_data_category_enum.dart';
import 'package:capacity_club_mobile_app/domain-features/domain/exercise-training/exercise-data/entity/exercise_training_entity.dart';

class ExerciseDataEntity {
  final UniqueId exercise_data_id;
  final String title;
  final List<String> description;
  final List<ExerciseDataCategory> categories;
  final String video_youtube_link;
  final String video_asset_link;
  final List<ExerciseTrainingEntity> exercise_training_list;
  final bool is_empty;

  ExerciseDataEntity(
      {required this.exercise_data_id,
      required this.title,
      required this.description,
      required this.categories,
      required this.video_youtube_link,
      required this.video_asset_link,
      required this.exercise_training_list,
      required this.is_empty});

  factory ExerciseDataEntity.empty() {
    return ExerciseDataEntity(
        exercise_data_id: UniqueId(''),
        title: '',
        description: [],
        categories: [],
        video_youtube_link: '',
        video_asset_link: '',
        exercise_training_list: [],
        is_empty: true);
  }
}
